src/field_theory/adjoin.lean

```diff
@@ -0,0 +1,177 @@
+/-
+Copyright (c) 2020 Thomas Browning and Patrick Lutz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning and Patrick Lutz
+-/
+
+import field_theory.subfield
+import field_theory.separable
+import field_theory.tower
+import group_theory.subgroup
+import field_theory.minimal_polynomial
+import linear_algebra.dimension
+import linear_algebra.finite_dimensional
+import ring_theory.adjoin_root
+import data.zmod.basic
+
+/-!
+# Adjoining Elements to Fields
+
+In this file we introduce the notion of adjoining elements to fields.
+This isn't quite the same as adjoining elements to rings.
+For example, algebra.adjoin K {x} might not include x⁻¹.
+
+## Main results
+
+(This is just a start, We've got more to add, including a proof of the Primitive Element Theorem)
+
+- `adjoin_twice`: adjoining S and then T is the same as adjoining S ∪ T.
+
+## Notation
+
+ - `F[α]` : Adjoin a single element α to F.
+-/
+
+variables (F : Type*) [field F] {E : Type*} [field E] [algebra F E] (S : set E)
+
+/--
+`adjoin F S` extends a field `F` by adjoining a set `S ⊆ E`.
+-/
+def adjoin : subalgebra F E := {
+  carrier := {
+    carrier := field.closure (set.range (algebra_map F E) ∪ S),
+    one_mem' := is_submonoid.one_mem,
+    mul_mem' := λ x y, is_submonoid.mul_mem,
+    zero_mem' := is_add_submonoid.zero_mem,
+    add_mem' := λ x y, is_add_submonoid.add_mem,
+  },
+  algebra_map_mem' := λ x, field.mem_closure (or.inl (set.mem_range.mpr ⟨x,rfl⟩)),
+}
+
+lemma adjoin.field_mem (x : F) : algebra_map F E x ∈ adjoin F S :=
+field.mem_closure (or.inl (set.mem_range_self x))
+
+lemma adjoin.field_subset : set.range (algebra_map F E) ⊆ adjoin F S :=
+begin
+  intros x hx,
+  cases hx with f hf,
+  rw ←hf,
+  exact adjoin.field_mem F S f,
+end
+
+instance adjoin.field_coe : has_coe_t F (adjoin F S) :=
+{coe := λ x, ⟨algebra_map F E x, adjoin.field_mem F S x⟩}
+
+lemma adjoin.set_mem (x : S) : ↑x ∈ adjoin F S :=
+field.mem_closure (or.inr (subtype.mem x))
+
+lemma adjoin.set_subset : S ⊆ adjoin F S :=
+λ x hx, adjoin.set_mem F S ⟨x,hx⟩
+
+instance adjoin.set_coe : has_coe_t S (adjoin F S) :=
+{coe := λ x, ⟨↑x, adjoin.set_mem F S x⟩}
+
+lemma adjoin.mono (T : set E) (h : S ⊆ T) : (adjoin F S : set E) ⊆ adjoin F T :=
+field.closure_mono (set.union_subset (set.subset_union_left _ _) (set.subset_union_of_subset_right h _))
+
+instance adjoin.is_subfield : is_subfield (adjoin F S : set E) := field.closure.is_subfield
+
+lemma adjoin_contains_field_as_subfield (F : set E) {HF : is_subfield F} : F ⊆ adjoin F S :=
+λ x hx, adjoin.field_mem F S ⟨x, hx⟩
+
+lemma adjoin_contains_subset {T : set E} {H : T ⊆ S} : T ⊆ adjoin F S :=
+begin
+  intros x hx,
+  exact adjoin.set_mem F S ⟨x,H hx⟩,
+end
+
+/-- proves an obvious linear equivalence (occasionally useful when working with dimension) -/
+definition adjoin_as_submodule_equiv : (adjoin F S) ≃ₗ[F] (adjoin F S) := {
```

Comment:
I'm confused... Is this just proving that the `X` is linearly equivalent to `X`, or is there some hidden coercion going on? Note that there is `linear_equiv.refl`.

Reply:
oh whoops. it used to be saying that adjoin F S is linearly equivalent to "the version of adjoin F S that is an F-submodule of E", but now that adjoin F S is an F-subalgebra of E, this is a tautology.